csharp/autobuilder/Semmle.Autobuild.Shared/Autobuilder.cs

```diff
@@ -283,21 +312,98 @@ void exitCallback(int ret, string msg, bool silent)
                 Log(silent ? Severity.Debug : Severity.Info, $"Exit code {ret}{(string.IsNullOrEmpty(msg) ? "" : $": {msg}")}");
             }
 
-            return script.Run(Actions, startCallback, exitCallback);
+            var onOutput = BuildOutputHandler(Console.Out);
+            var onError = BuildOutputHandler(Console.Error);
+
+            var buildResult = script.Run(Actions, startCallback, exitCallback, onOutput, onError);
+
+            // if the build succeeded, all diagnostics we captured from the build output should be warnings;
+            // otherwise they should all be errors
+            var diagSeverity = buildResult == 0 ? DiagnosticMessage.TspSeverity.Warning : DiagnosticMessage.TspSeverity.Error;
+            foreach (var diagResult in this.DiagnosticClassifier.Results)
+            {
+                var diag = diagResult.ToDiagnosticMessage(this);
+                diag.Severity = diagSeverity;
+
+                Diagnostic(diag);
+            }
+
+            return buildResult;
         }
 
         /// <summary>
         /// Returns the build script to use for this project.
         /// </summary>
         public abstract BuildScript GetBuildScript();
 
+        /// <summary>
+        /// Constructs a standard <see cref="DiagnosticMessage" /> for some message with
+        /// <see cref="id" /> and a human-friendly <see cref="name" />.
+        /// </summary>
+        /// <param name="id">The last part of the message id.</param>
+        /// <param name="name">The human-friendly description of the message.</param>
+        /// <returns>The resulting <see cref="DiagnosticMessage" />.</returns>
+        public DiagnosticMessage MakeDiagnostic(string id, string name)
```

Comment:
I think I will probably want to keep `Options` out of `DiagnosticMessage`, in case we need to emit diagnostics from the extractor as well, rather than just the autobuilder. 

I have made some changes here though to be able to set more fields directly through the respective constructors as well.